### PR TITLE
Fix mismatch C/AVX2 kernel svt_av1_apply_temporal_filter_planewise()

### PR DESCRIPTION
**Pull request by spawlows**
# Description
Calculation double/float in some cases can return different value.

# Issue
#1428 

# Author(s)
Slawomir Pawlowski

# Performance impact

- [x] quality
- [ ] memory
- [x] speed
- [x] 8 bit
- [x] 10 bit
- [ ] N/A

# Test set
- [ ] obj-1-fast can be found [here](https://media.xiph.org/video/derf/objective-1-fast.tar.gz)
- [x] other
- [ ] N/A

# Merge method
- [x] Allow the maintainer to squash and merge when PR is ready to create a 1-commit to the master branch. The maintainer will be able to fix typos / combine commit messages to create a more readable 1-commit message or use whatever is stated in the 'Description' section
- [ ] I will clean up my commits and the maintainer shall use 'rebase and merge' to the master branch

### CODE REVIEW

**Comment by hassount** [comment]
LGTM, Pending review from @ttrigui  and testing results from @OmranAbdelkafi

**Comment by ttrigui** [comment]
My understanding was that we decided to use float for the c code instead of double to match the AVX2 code. Is there a reason why we are also changing the AVX2 functions?

**Comment by spawlows** [comment]
Hi @ttrigui 
I modified AVX2 kernel to use double variables in some places because there are some mathematical tricks like multiply by inverse of digit instead of division(this kind of operation can return different results in some cases when float is used), but changes in C are still needed to match bitstreams because expf()[float] is less expensive than exp()[double].
  
Compared to results that I shared with you previously there is smaller quality drop with similar performance.
 
I also found new optimization that reduce AVX2 implementation time from about 2.1% to 0.9% (please look at line 263 or 540; there was power function that uses float when input was on integers).

To summarize this changes in C and AVX2 produce less quality drop and is faster.

**Comment by ttrigui** [comment]
> Hi @ttrigui
> I modified AVX2 kernel to use double variables in some places because there are some mathematical tricks like multiply by inverse of digit instead of division(this kind of operation can return different results in some cases when float is used), but changes in C are still needed to match bitstreams because expf()[float] is less expensive than exp()[double].
> 
> Compared to results that I shared with you previously there is smaller quality drop with similar performance.
> 
> I also found new optimization that reduce AVX2 implementation time from about 2.1% to 0.9% (please look at line 263 or 540; there was power function that uses float when input was on integers).
> 
> To summarize this changes in C and AVX2 produce less quality drop and is faster.

I see. So to fix the mismatch both C and AVX2 code had to be modified.
I'm glad that you changed the powf. I did profiling and I noticed that it is consuming many cycles.
We will then have to test AVX2 before and after PR to evaluate the impact, C before and after PR to evaluate the impact going from double to float and C vs AVX2 after the PR to make sure we do not have any more mismatches.
@OmranAbdelkafi does the testing make sense to you?

**Comment by paulsgh** [comment]
Tested objective 1 fast 8bit and 10bit:

Note: this PR is not lossless.
**M4 - 10bit: BDRate** = 0.002825%
**M8 - 10bit: BDRate** = 0.003405%

� | Speed Deviation  | BD-Rate   Deviation | Memory Deviation
-- | -- | -- | --
M4 - 8bit| 0.14% | -0.01% | -0.01%
M8 - 8bit| 0.20% | -0.01% | 0.00%
M4 - 10bit| 0.03% | **0.00%*** | -0.01%
M8 - 10bit| 0.11% | **0.00%*** | 0.00%

**Comment by paulsgh** [comment]
Tested objective 1 fast 8bit and 10bit:
Note: this PR is not lossless.

� | Speed Deviation  | BD-Rate   Deviation | Memory Deviation
-- | -- | -- | --
M0 - 8bit| 0.59% | 0.00365% | 0.00%
M0 - 10bit| 0.31% | -0.00275% | -0.01%

**Comment by hassount** [comment]
@ttrigui / @OmranAbdelkafi is the BDR difference is as expected ?

**Comment by OmranAbdelkafi** [comment]
Compared avx12 vs c using obj-1-fast 8bit test-set within presets 4 and 8 : Lossless comparison. @hassount Approved PR on my side.

**Comment by hassount** [comment]
@ttrigui @OmranAbdelkafi thank you both, merging.
